pkg/restore/actions/csi/pvc_vs_dv_action.go

```diff
@@ -0,0 +1,137 @@
+func (p *pvcVSDVRestoreItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"persistentvolumeclaims"},
+
+	}, nil
+}
+
+
+// Execute modifies the PVC's spec to use the VolumeSnapshot object as the
+// data source ensuring that the newly provisioned volume can be pre-populated
+// with data from the VolumeSnapshot.
+func (p *pvcVSDVRestoreItemAction) Execute(
+	input *velero.RestoreItemActionExecuteInput,
+) (*velero.RestoreItemActionExecuteOutput, error) {
+	var pvcFromBackup corev1api.PersistentVolumeClaim
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(
+		input.ItemFromBackup.UnstructuredContent(), &pvcFromBackup); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	logger := p.log.WithFields(logrus.Fields{
+		"Action":  "PVCVSDVRestoreItemAction",
+		"PVC":     pvcFromBackup.Namespace + "/" + pvcFromBackup.Name,
+		"Restore": input.Restore.Namespace + "/" + input.Restore.Name,
+	})
+	logger.Info("Starting PVCVSDVRestoreItemAction for PVC")
+
+	var pvc corev1api.PersistentVolumeClaim
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(input.Item.UnstructuredContent(), &pvc); err != nil {
+		return nil, errors.Wrap(err, "failed to convert PVC from unstructured")
+	}
+
+	additionalItems := []velero.ResourceIdentifier{}
+
+	// Check for VolumeSnapshot annotation.
+	// Velero uses velerov1api.VolumeSnapshotLabel to store the snapshot name.
+	if pvc.Annotations != nil {
+		if vsName, ok := pvc.Annotations[velerov1api.VolumeSnapshotLabel]; ok && vsName != "" {
+			additionalItems = append(additionalItems, velero.ResourceIdentifier{
+				GroupResource: kuberesource.VolumeSnapshots,
+				Name:          vsName,
+				Namespace:     pvc.Namespace,
+			})
+		}
+
+		// Check for DataVolume annotation.
+		if dvName, ok := pvc.Annotations[velerov1api.DataUploadNameAnnotation]; ok && dvName != "" {
+			additionalItems = append(additionalItems, velero.ResourceIdentifier{
+				GroupResource: schema.GroupResource{
+					Group:    "cdi.kubevirt.io",
+					Resource: "datavolumes",
+				},
+				Name:      dvName,
+				Namespace: pvc.Namespace,
+			})
+		}
+	}
+
+	return &velero.RestoreItemActionExecuteOutput{
+		UpdatedItem:     input.Item,
+		AdditionalItems: additionalItems,
+	}, nil
+}
+
+
+
+func (p *pvcVSDVRestoreItemAction) Name() string {
+	return "PVCRestoreItemAction"
+}
+
+
+func (p *pvcVSDVRestoreItemAction) Cancel(
+	operationID string, restore *velerov1api.Restore) error {
+	return nil
+}
+
+func (p *pvcVSDVRestoreItemAction) AreAdditionalItemsReady(
+	additionalItems []velero.ResourceIdentifier,
+	restore *velerov1api.Restore,
+) (bool, error) {
+	return true, nil
+}
+
+// NewPvcVSDVRestoreItemAction returns a new instance of PVCVSDVRestoreItemAction.
+func NewPvcVSDVRestoreItemAction(f client.Factory) plugincommon.HandlerInitializer {
+	return func(logger logrus.FieldLogger) (any, error) {
+		crClient, err := f.KubebuilderClient()
```

Comment:
It looks like the plugin doesn't actually need crClient, so this can probably be avoided.